### ISSUE TITLE
@ConfigurationProperties should have an option to have properties read each time from Environment 

### ISSUE THREAD

**Issue by philwebb** [spring-projects/spring-boot#259]
Originally raised on the SPR issue tracker:
https://jira.springsource.org/browse/SPR-11289

`@ConfigurationProperties` objects should have some mechanism for allowing some or all properties to be read from a propertySource on each get invocation. It seems like instead of injected a concrete implementation into the AppContext we should inject a proxy for the Config object which can reference the Environment abstraction.

The usecase for this would be when we have implemented a custom PropertySource which is aware of property changes.

**Comment by rachelwilson** [comment]
I was wondering if someone could point me (and other readers) in the direction of the "recent announcement about dynamic Spring props" as mentioned by andrewharmellaw above.  I was interested to hear what the latest thinking is and I cant seem to find anything recent.

**Comment by dsyer** [comment]
Try here: http://projects.spring.io/spring-cloud/spring-cloud.html#_environment_changes (and some other bits of that section).

**Comment by rachelwilson** [comment]
Hmm, that seems to be documentation rather than an announcement as such.  Can I assume in that case that the dynamic properties apply only to the Cloud and Boot products and there are no announced changes to the original Spring framework?

**Comment by dsyer** [comment]
Yes, it's a Cloud feature right now (but with really minimal dependencies, as long as you are using Spring Boot).

**Comment by nmarasoiu** [comment]
Hi, I am having this issue with Spring Cloud: I added a MapPropertySource to the env, I am changing the values from the ConcurentMap from another thread, but I can't see those values reflecting in env.getProperty or in @ConfigurationProperties beans. Pls advise!

**Comment by dsyer** [comment]
If you really changed the environment, you would see it in `getProperty()`, so if you didn't see it there, then I suspect you didn't change it, or you changed a property that is overridden in another property sourec. You can see all the property sources in the /env endpoint.

You have to rebind to get `@ConfigurationProperties` to update, so that's a different problem. There's a `ContextRefresher` in spring-cloud-context that you can use if you want to trigger it manually.

**Comment by wilkinsona** [comment]
We've discussed the possibility of supporting a property that's a `Supplier<?>` that retrieves the value from the `Environment` each time `get()` is called. Our feeling is that this isn't a viable solution for two reasons:

1. For a frequently used infrequently changed piece of configuration, retrieving the property from the environment each time is inefficient
2. Retrieving a consistent view of updates to multiple properties is hard, if not impossible.

Our feeling, as it [was above](
https://github.com/spring-projects/spring-boot/issues/259#issuecomment-33159282), is that an event-based approach would be better. The event would provide a complete new instance of a configuration properties class with the changed value or values. This addresses both of the problems described above while also working with immutable configuration properties that are currently in plan for 2.2.

The question remains if this is a good idea. We'd also be interested in hearing some of the use cases that a change in this area would enable. As things stand, there hasn't been particularly broad interest in this functionality. Given the reasonably large amount of complexity that it would introduce, it's difficult to justify implementing it at this time.

**Comment by Code2Life** [comment]
Already implemented in my project: [https://github.com/Code2Life/spring-boot-dynamic-config](https://github.com/Code2Life/spring-boot-dynamic-config)

Adding @DynamicConfig on @ConfigurationProperties simply works.

This lib is widely used in the company I work for. Hope this feature becomes a built-in feature of Spring Boot.  @philwebb @dsyer

**Comment by denon82** [comment]
Great job. As of a suggestion I would remove the Lombok dependency and only
depend on spring codebase.

One question: this dynamic configuration only affects the given project?
What about share configuration between project? That what spring-config is
intended to do, while your project is intented to make the configurations
dynamic to the project code base. How about security?

Do you want to incorporate your codebase to spring's codebase? Of course,
if they consider it interesting.

Regards,

Flavio Oliva

**Comment by Code2Life** [comment]
@denon82 Thanks ! 
1. this dynamic config only affects given project because it's enhancement of spring boot config native ways.
2. config sharing could be on orchestration layer, for eg. mount same ConfigMap as files in different Kubernetes workload
3. as for security, there are conditional beans, so the config file will not be dynamic unless the owner intends to. Moreover, the source of config files are in other systems like Git, thus, the security responsibilities are shifted to Git permission, CI/CD pipeline security, Kubernetes RBAC, etc.

**Comment by philwebb** [comment]
Thanks very much for sharing the project @Code2Life. It looks like an interesting alternative to `@RefreshScope` for those that want it. I don't have any specific comments on the code (although I agree with @denon82 about Lombok). We're unlikely to get to this feature in Spring Boot itself for a little while so it's great that users have another option.

**Comment by Code2Life** [comment]
Thanks for reviewing @philwebb , `@RefreshScope` relies on spring cloud stack and additional config server. we've tried spring-cloud-config and other spring-cloud components in production years ago and found it's too heavyweight, then we went back to Spring Boot.

**Comment by dsyer** [comment]
FYI `spring-cloud-config` isn't needed for `@RefreshScope`, and also `@RefreshScope` is not needed for `@ConfigurationProperties`. Just put `spring-cloud-commons` on the classpath - it's very lightweight, really - the only dependency is Spring Boot.

**Comment by shiyiyue1102** [comment]
@RefreshScope may not be secure enough as it will first destroy the current bean and then create a new bean. What if the environment properties encounter an exception when creating a new bean? This online running application will be a disaster. Is it possible to adjust the order and first try creating a temporary bean? If it can be initialized normally, then replace the original bean. This will be a significant optimization for runtime configuration updates.

**Comment by shiyiyue1102** [comment]
In addition，if it is possible to move @RefreshScope from spring-cloud to spring boot?  it should not be bundled to spring cloud project.  

I am a core builder in the Nacos community（https://nacos.io/en/） which is widely used by chinese  developer community and I am responsible for config center. 
We want to implement all configurations on the hosted application side, support elegant updates during runtime, and improve the efficiency of business software iteration. If Spring can support elegant refresh beans, then there is no need for every second party middleware to implement this feature

**Comment by wilkinsona** [comment]
There are no plans to move `@RefreshScope` to Spring Boot. If you have concerns about its behavior, please raise a Spring Cloud Commons issue.
